sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresInt2Tests.java

```diff
@@ -0,0 +1,397 @@
+                        " -0.5 |         -1\n" +
+                        "  0.0 |          0\n" +
+                        "  0.5 |          1\n" +
+                        "  1.5 |          2\n" +
+                        "  2.5 |          3"
+        );
+    }
+
+    @Test @Ignore("non decimals not suppported yet")
+    public void testNonDecimals() {
+        this.qs(
+                "SELECT '0b100101'::INT2 as x;\n" +
+                        " x \n" +
+                        "---\n" +
+                        " 37\n" +
+                        "(1 row)\n" +
+                        "\n" +
+                        "SELECT '0o273'::INT2;\n" +
+                        " x \n" +
+                        "---\n" +
+                        " 187\n" +
+                        "(1 row)\n" +
+                        "\n" +
+                        "SELECT '0x42F'::INT2;\n" +
+                        " x \n" +
+                        "---\n" +
+                        " 1071\n" +
+                        "(1 row)\n" +
+                        "\n"
+        );
+    }
+
+    @Test @Ignore("unimplemented operator")
+    public void testLeftShift() {
+        this.qs(
+        "SELECT (-1::INT2 << 15)::TEXT;\n" +
+                "  text  \n" +
+                "--------\n" +
+                " -32768\n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT ((-1::int2<<15)+1::int2)::text;\n" +
+                "  text  \n" +
+                "--------\n" +
+                " -32767\n" +
+                "(1 row)"
+        );
+    }
+
+    // Ignored because this fails in Postgres but here we get:
+    @Test @Ignore
+    public void testSelectOverflow() {
+        String error_message = "INT2 out of range";
+        // We get:
+        // L: (Some(-32767), Some(2))x1 --> wraps around
+        // L: (Some(-1234), Some(-2468))x1
+        // L: (Some(0), Some(0))x1
+        // L: (Some(1234), Some(2468))x1
+        // L: (Some(32767), Some(-2))x1 --> wraps around
+        // Taken from: https://github.com/postgres/postgres/blob/c161ab74f76af8e0f3c6b349438525ad9575683b/src/test/regress/expected/int2.out#L202C1-L203C30
+        this.shouldFail("SELECT i.f1, i.f1 * 2::INT2 AS x FROM INT2_TBL i", error_message);
+
+        // We get:
+        // L: (Some(-32767), Some(-32765))x1
+        // L: (Some(-1234), Some(-1232))x1
+        // L: (Some(0), Some(2))x1
+        // L: (Some(1234), Some(1236))x1
+        // L: (Some(32767), Some(-32767))x1 --> wraps around
+        // https://github.com/postgres/postgres/blob/c161ab74f76af8e0f3c6b349438525ad9575683b/src/test/regress/expected/int2.out#L223
+        this.shouldFail( "SELECT i.f1, i.f1 + '2'::INT2 AS x FROM INT2_TBL i", error_message);
+
+        // Similarly,
+        this.shouldFail("SELECT i.f1, i.f1 - '2'::INT2 AS x FROM INT2_TBL i;", error_message);
+    }
+
+    // This passes for the Calcite version but fails for the run time version
```

Comment:
I expect there must exist a Rust compiler flag to silence these errors.
We should give warnings ourselves about this probably.
Please file an issue to create an analysis which looks for such problems at compilation time.

Reply:
We can have `#[allow(unconditional_panic)]` to silence the error, but it will panic during run time. Which is not consistent with the Calcite optimized version. I think we should manually check for modulo with `abs(right) = 1` and return 0.

Reply:
If Calcite does something here which is not an error it's wrong.
This happens because integers in two's complement can represent one more negative value than positive values. For example, a byte can be from -128 to 127. So abs(-128) has no representation. The correct action is to give an error when this happens.

Reply:
I have added the unconditional_panic flag to PR #1183